server/src/main/java/org/elasticsearch/action/bulk/BulkShardRequestInferenceProvider.java

```diff
@@ -0,0 +1,296 @@
+                    @Override
+                    public void onResponse(InferenceServiceResults results) {
+
+                        if (results == null) {
+                            throw new IllegalArgumentException(
+                                "No inference retrieved for model ID " + modelId + " in document " + docWriteRequest.id()
+                            );
+                        }
+
+                        int i = 0;
+                        for (InferenceResults inferenceResults : results.transformToLegacyFormat()) {
+                            String fieldName = inferenceFieldNames.get(i++);
+                            @SuppressWarnings("unchecked")
+                            List<Map<String, Object>> inferenceFieldResultList = (List<Map<String, Object>>) rootInferenceFieldMap
+                                .computeIfAbsent(fieldName, k -> new ArrayList<>());
+                            // Remove previous inference results if any
+                            inferenceFieldResultList.clear();
+
+                            // TODO Check inference result type to change subfield name
+                            var inferenceFieldMap = Map.of(
+                                SPARSE_VECTOR_SUBFIELD_NAME,
+                                inferenceResults.asMap("output").get("output"),
+                                TEXT_SUBFIELD_NAME,
+                                docMap.get(fieldName)
+                            );
+                            inferenceFieldResultList.add(inferenceFieldMap);
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        failedItems.add(itemIndex);
+                        onBulkItemFailure.accept(bulkItemRequest, e);
+                    }
+                };
+                inferenceProvider.service()
+                    .infer(
+                        inferenceProvider.model,
+                        inferenceFieldNames.stream().map(docMap::get).map(String::valueOf).collect(Collectors.toList()),
+                        // TODO check for additional settings needed
+                        Map.of(),
+                        InputType.INGEST,
+                        ActionListener.releaseAfter(inferenceResultsListener, docRef.acquire())
+                    );
+            }
+        }
+    }
+
+    private static List<String> getFieldNamesForInference(Map.Entry<String, Set<String>> fieldModelsEntrySet, Map<String, Object> docMap) {
+        List<String> inferenceFieldNames = new ArrayList<>();
+        for (String inferenceField : fieldModelsEntrySet.getValue()) {
+            Object fieldValue = docMap.get(inferenceField);
+
+            // Perform inference on string, non-null values
+            if (fieldValue instanceof String) {
+                inferenceFieldNames.add(inferenceField);
+            }
```

Comment:
Do we need to handle when the field value is a non-null & non-String value (i.e. when the user has provided a value with an invalid data type)? Or will that be handled somewhere downstream/upstream?

Reply:
Good catch, we don't as of now. 

There are some cases to consider, I'll work on them
- Array values: We could treat these as chunking, and perform inference on every array value
- Non-string values: Convert them to strings before doing inference. Don't error out when we have a non-string value, similar to how text works.

My only concern on converting non-strings is that we're doing inference on fields where inference makes no sense - and potentially incurring in costs - instead of warning the user.

Reply:
I wasn't even thinking of multi-valued text fields (i.e. array of strings), but that's a case we need to handle here as well.

I was thinking of handling obvious error cases, such as when the value is a `Map` and can't be converted to a string in a sensible way.

Reply:
Just checked how the `text` field handles this:

Primitive value (i.e. string, bool, number): Coerce to string
Array of primitive values: Index each value separately, coerce each value to string
Object (i.e. `Map`): Throw error
Array containing an object: Throw error